docs/architecture/psa-migration/md-cipher-dispatch.md

```diff
@@ -0,0 +1,450 @@
+
+A hash is available through the mixed-domain interface iff either of the following conditions is true:
+
+* A legacy hash interface is available and the hash algorithm is implemented in software.
+* PSA crypto is enabled and the hash algorithm is implemented via PSA.
+
+We could go further and make PSA accelerators available to legacy callers that call any legacy hash interface, e.g. `md.h` or `shaX.h`. There is little point in doing this, however: callers should just use the mixed-domain interface.
+
+#### Shape of the mixed-domain hash interface
+
+We now need to create an abstraction for mixed-domain hash calculation. (We could not create an abstraction, but that would require every piece of mixed-domain code to replicate the logic here. We went that route in Mbed TLS 3.3, but it made it effectively impossible to get something that works correctly.)
+
+Requirements: given a hash algorithm,
+
+* Obtain some metadata about it (size, block size).
+* Calculate the hash.
+* Set up a multipart operation to calculate the hash. The operation must support update, finish, reset, abort, clone.
+
+The existing interface in `md.h` is close to what we want, but not perfect. What's wrong with it?
+
+* It has an extra step of converting from `mbedtls_md_type_t` to `const mbedtls_md_info_t *`.
+* It includes extra fluff such as names and HMAC. This costs code size.
+* The md module has some legacy baggage dating from when it was more open, which we don't care about anymore. This may cost code size.
+
+These problems are easily solvable.
+
+* `mbedtls_md_info_t` can become a very thin type. We can't remove the extra function call from the source code of callers, but we can make it a very thin abstraction that compilers can often optimize.
+* We can make names and HMAC optional. The mixed-domain hash interface won't be the full `MBEDTLS_MD_C` but a subset.
+* We can optimize `md.c` without making API changes to `md.h`.
+
+## Specification
+
+### MD light
+
+https://github.com/Mbed-TLS/mbedtls/pull/6474 implements part of this specification, but it's based on Mbed TLS 3.2, so it needs to be rewritten for 3.3.
+
+#### Definition of MD light
+
+MD light is a subset of `md.h` that implements the hash calculation interface described in ”[Designing an interface for hashes](#designing-an-interface-for-hashes)”. It is activated by `MBEDTLS_MD_LIGHT` in `mbedtls_config.h`.
+
+The following things enable MD light automatically in `build_info.h`:
+
+* A [mixed-domain](#classification-of-callers) module that needs to calculate hashes is enabled.
+* `MBEDTLS_MD_C` is enabled.
+
+MD light includes the following types:
+
+* `mbedtls_md_type_t`
+* `mbedtls_md_info_t`
+* `mbedtls_md_context_t`
+
+MD light includes the following functions:
+
+* `mbedtls_md_info_from_type`
+* `mbedtls_md_init`
+* `mbedtls_md_free`
+* `mbedtls_md_setup` — but `hmac` must be 0 if `MBEDTLS_MD_C` is disabled.
+* `mbedtls_md_clone`
```

Comment:
What's the reason for including `clone` here? A quick grep suggests it's only used in TLS, which I think can keep depending on the full MD.

Reply:
Because it's used in TLS and any I think any existing mixed-domain code should be able to run on md-light (as long as it doesn't need HMAC).

Reply:
Currently TLS depends on `MD_C || USE_PSA_CRYPTO`. So just to check I understand correctly, you're suggesting we change that from `MD_C` to `MD_LIGHT`. Then in `ssl_ciphersuites.c` all the ciphersuites using HMAC (unless I'm mistaken that's just the CBC and NULL ciphersuites these days) will need to declare a dependency on `MD_C || USE_PSA_CRYPTO` - on top of the dependency on the specific hash being used.

That's certainly doable, but what would be the upside compared to the status quo?

Reply:
This part is not about HMAC. Supporting clone has minimal cost (it just follows the same mechanics as the other functions), I think it's simpler to make the same changes than to have an exception in the design.

Reply:
To be sure, when you say "an exception in the design", you're saying that `MD_LIGHT` covers everything that's not HMAC, and `md_clone()` is not about HMAC, so excluding it from `MD_LIGHT` would be the exception that you'd rather avoid?

Or are you referring to something else?

Reply:
Yes: `MD_LIGHT` covers all md operations. `MD_C` = `MD_LIGHT` + HMAC + metadata.

Reply:
Ok, I agree that's a good argument for keeping `md_clone()` in `MD_LIGHT`, so let's do that.

Reply:
> Because it's used in TLS and any I think any existing mixed-domain code should be able to run on md-light (as long as it doesn't need HMAC).

For the record, TLS needs HMAC: the TLS 1.2 PRF is based on HMAC (implemented using `mbedtls_md_hmac_xxx()` functions when `USE_PSA_CRYPTO` is not defined), and TLS 1.3 uses HKDF which is also based on HMAC.

So, TLS will keep depending on the full `MD_C` (or `USE_PSA_CRYPTO`), not just `MD_LIGHT`.

Reply:
TLS 1.3 uses HKDF through a PSA API, which does not involve md. But indeed TLS 1.2 without `MBEDTLS_USE_PSA_CRYPTO` needs the full MD anyway, at least until we have an alternative for HMAC.